runtime/common/marshaling/index.ts

```diff
@@ -1,106 +1,154 @@
+//     {type: "Error", path: ['errorField']}
+//   ]
+// }
+//
+// When we pass these objects into or out of isolated-vm we also need to set "copy: true" to enable
+// the structuredClone() algorithm.
 
 const MaxTraverseDepth = 100;
 
-const customMarshalers = [marshalError, marshalBuffer, marshalNumber, marshalDate];
-
-const customUnmarshalers: Array<(val: any) => any> = [unmarshalError, unmarshalBuffer, unmarshalNumber, unmarshalDate];
-
-function serialize(val: any): any {
-  for (const marshaler of customMarshalers) {
-    const result = marshaler(val);
-    if (result !== undefined) {
-      return result;
-    }
-  }
-
-  return val;
-}
-
-function deserialize(_: string, val: any): any {
-  if (val) {
-    for (const unmarshaler of customUnmarshalers) {
-      const result = unmarshaler(val);
-      if (result !== undefined) {
-        return result;
-      }
-    }
-  }
-
-  return val;
-}
-
-function processValue(val: any, depth: number = 0): any {
+// pathPrefix can be temporarily modified, but needs to be restored to its original value
+// before returning.
+//
+// "hasModifications" is to avoid trying to copy objects that don't need to be copied.
+// Only objects containing a buffer or an error should need to be copied.
+function fixUncopyableTypes(
+  val: any,
+  pathPrefix: string[],
+  postTransforms: object[],
+  depth: number = 0,
+): {val: any; hasModifications: boolean} {
   if (depth >= MaxTraverseDepth) {
     // this is either a circular reference or a super nested value that we mostly likely
     // don't care about marshalling.
-    return val;
+    return {val, hasModifications: false};
   }
 
-  if (val === undefined) {
-    return HACK_UNDEFINED_JSON_VALUE;
+  if (!val) {
+    return {val, hasModifications: false};
   }
 
-  if (val === null) {
-    return val;
+  const maybeError = marshalError(val);
+  if (maybeError) {
+    postTransforms.push({
+      type: 'Error',
+      path: [...pathPrefix],
+    });
+    return {val: maybeError, hasModifications: true};
   }
 
-  if (Array.isArray(val)) {
-    return val.map(item => processValue(item, depth + 1));
+  if (val instanceof Buffer || global.Buffer?.isBuffer(val)) {
+    // Theoretically it should be possible to pass an array buffer
+    // through structured copy with some transfer options, but it's
+    // simpler to just encode it as a string.
+    postTransforms.push({
+      type: 'Buffer',
+      path: [...pathPrefix],
+    });
+    return {val: val.toString('base64'), hasModifications: true};
   }
 
-  const serializedValue = serialize(val);
-
-  if (typeof serializedValue === 'object') {
-    let objectValue = serializedValue;
-    if ('toJSON' in serializedValue && typeof serializedValue.toJSON === 'function') {
-      objectValue = serializedValue.toJSON();
+  if (Array.isArray(val)) {
+    const maybeModifiedArray: any[] = [];
+    let someItemHadModifications = false;
+    val.forEach((item, index) => {
```

Comment:
Super nit (feel free to ignore): prefer `for` loop over `forEach` in modern js due to performance difference.

Reply:
Done